### ISSUE TITLE
Value Checker ignores annotations on casts

### ISSUE THREAD

**Issue by panacekcz** [typetools/checker-framework#1333]
I realize this is not very important because users would rarely write annotations on casts, but it is quite unexpected behavior.

```java
import org.checkerframework.common.value.qual.IntVal;
public class CastReport {
 void intcasts(@IntVal(1) int i){
  @IntVal(1) int z1 = (@IntVal(2) int)i;
  @IntVal(2) int z2 = (@IntVal(2) int)i;
 }
}
```
The Value checker ignores the annotation on the cast and keeps the qualifier of the operand, giving this warning:
```
CastReport.java:5: error: [assignment.type.incompatible] incompatible types in assignment.
		@IntVal(2) int z2 = (@IntVal(2) int)i;
		                    ^
  found   : @IntVal(1) int
  required: @IntVal(2L) int
1 error
```

I think one of the following ways would be better than the current situation:

1. Use the annotation on the cast and ignore the qualifier of the operand. Warn about unsafe cast if the qualifier on the cast is not a subtype of the qualifier of the operand.
2. Use the greatest lower bound of the qualifiers from the cast and the operand. This would give the best precision. Again, warn about unsafe cast if the qualifier on the cast is not a subtype of the qualifier of the operand.

(I am interested in what is the best way to handle that because I need to modify the implementation of casting in the value checker to correctly handle `@ArrayLen` on strings, and I would rather not implement the wrong logic.)

**Comment by smillst** [comment]
Thanks for pointing this out.
>Use the annotation on the cast and ignore the qualifier of the operand. Warn about unsafe cast if the qualifier on the cast is not a subtype of the qualifier of the operand.

This is the correct behavior.

Looking at the code, [this call to `replaceAnnotation`](https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java#L948) should be changed to `AnnotatedTypeMirror#addMissingAnnotations` and so should the other call to `replaceAnnotation` in this method.  Would you mind fixing this?

**Comment by panacekcz** [comment]
Ok, thank you for the explanation, I will do that.
